packages/start/vite/plugin.js

```diff
@@ -671,44 +672,48 @@ export default function solidStart(options) {
   ].filter(Boolean);
 }
 
+/**
+ * @returns {import('vite').Plugin}
+ */
 function islands() {
   return {
     name: "solid-start-islands",
-    load(id) {
-      if (id.endsWith("?island")) {
-        return {
-          code: `
-            import Component from '${id.replace("?island", "")}';
+    async transform(code, id, opts) {
+      if (code.includes("unstable_island")) {
+        const ssr = opts?.ssr === true;
 
-            window._$HY.island("${id.slice(process.cwd().length)}", Component);
+        /** @type {Record<string, string>} */
+        const islandComponentIds = Object.fromEntries(
+          await Promise.all(
+            [
+              ...code.matchAll(
+                /const ([A-Za-z_]+) = unstable_island\(\(\) => import\((("([^"]+)")|('([^']+)'))\)\)/g
+              )
+            ].map(async ([a, b, c]) => {
+              c = c.slice(1, -1);
+              const resolved = await this.resolve(c, id);
+              if (!resolved) {
+                throw new Error(`Cannot resolve ${c} from ${id}`);
+              }
+              const componentId = createHash(resolved.id);
+              return [c, componentId];
+            })
+          )
+        );
 
-            export default Component;
-            `
-        };
-      }
-    },
-    /**
-     * @param {any} id
-     * @param {string} code
-     */
-    transform(code, id, ssr) {
```

Comment:
FYI this `ssr` flag was not working since Vite 2.7.

https://vitejs.dev/guide/ssr.html#:~:text=Before%20Vite%202.7%2C%20this%20was%20informed%20to%20plugin%20hooks%20with%20a%20positional%20ssr%20param%20instead%20of%20using%20the%20options%20object.%20All%20major%20frameworks%20and%20plugins%20are%20updated%20but%20you%20may%20find%20outdated%20posts%20using%20the%20previous%20API.